rfcs/rfc-story-based-git-workflow.md

```diff
@@ -0,0 +1,117 @@
+6. As needed, Dev1 and Dev2 compare their branch with other story branches that are currently in progress (same reason as above). 
+7. When Dev1 and Dev2 are satisfied that their story is complete and is up-to-date with MASTER, they create a pull request.
+8. The code is reviewed and a few small issues are corrected.
+9. The code is merged into master (squashing commits from the story branch) and the full test suite is re-run in the E2E environment. Because the tests have been run already, there is little chance of breaking the build at this point.
+10. The story branch is cleaned up (destroyed).
+
+# Motivation
+
+The original entities team ("A" team, "Coops" team, team "Aardvark") has been using an approach where they all fork the master branch individually and commit small changes as often as possible. The advantage to this approach is that it keeps merges incremental and small. 
+
+Adding team Bumblebee (FreshWorks) as codebase contributors to the <https://github.com/bcgov/lear> repository exposed some workflow issues around being able to continuously develop with non co-located teams comprised of many team members. The review and commit strategies that were effective with one team do not scale to two teams. There are some other advantages to story branching that can be realized as well.
+
+Here are some of the issues we have seen:
+
+* Frequent small commits that invalidate the test cycle for a story that is almost ready to be created as a pull request
+* Lack of transparency between developers' work in progress (individual developer forks live on GitHub accounts on an ad-hoc basis)
+* Commits to master that are incomplete or break the build
+* Commits to master that do not pass linting
+* Developers committing their own code without a review
+
+# Detailed design
+
+No detailed design necessary as the actual technical aspects are not complicated. This RFC describes a potential team agreement for WoW (way of working), not a technical design change.
+
+# Drawbacks
+
+- Disrupting the current team's WoW (Way of Working)
+- Not consistent with how public committers would be working (if we were using them - although we could create branches for them as well if the volume is low)
+- Currently difficult to do E2E testing in local development environment
+- Risk of "stories" being defined in a very course-grained way and loooooong running PRs being difficult/impossible to merge
+
+
+# Alternatives
+
+## Running tests in a PR environment before merging
+If an E2E environment can be set up in an automated fashion, webhooks can be used to trigger a dedicated environment set up when the PR is created. All the automated tests will be run in that environment. 
```

Comment:
Again this can be done with the current strategy, see projects such as Linux, OpenStack, Android ....

Reply:
Correct, this doesn't actually address the gitflow and could be done with the current strategy. However, it's overhead that a team probably wouldn't want to take on unless merges are of a certain size (waiting for a full environment and test cycle seems "heavy" if the merges are small and frequent).